conformance-tests/src/main.rs

```diff
@@ -0,0 +1,264 @@
+                test_environment::manifest_template::EnvTemplate::from_file(&test.manifest)
+                    .unwrap();
+            manifest.substitute(e, |_| None).unwrap();
+            e.write_file("spin.toml", manifest.contents())?;
+            e.copy_into(&test.component, test.component.file_name().unwrap())?;
+            Ok(())
+        })
+        .unwrap();
+        for invocation in test.config.invocations {
+            let conformance_tests::config::Invocation::Http(mut invocation) = invocation;
+            invocation.request.substitute_from_env(&mut env).unwrap();
+            let shim = env.runtime_mut();
+            if let Err(e) = invocation.run(|request| shim.make_http_request(request)) {
+                println!("❌ test failed: {}", test.name);
+                println!("error: {}", e);
+                for e in e.chain() {
+                    println!("\t{}", e);
+                }
+
+                continue 'test;
+            }
+        }
+        println!("✅ test passed: {}", test.name);
+    }
+}
+
+struct SpinShim {
+    process: std::process::Child,
+    #[allow(dead_code)]
+    stdout: OutputStream,
+    stderr: OutputStream,
+    io_mode: IoMode,
+}
+
+/// `ctr run` invocations require an ID that is unique to all currently running instances. Since
+/// only one test runs at a time, we can reuse a constant ID.
+const CTR_RUN_ID: &str = "run-id";
+
+impl SpinShim {
+    pub fn config(
+        ctr_binary: PathBuf,
+        spin_binary: PathBuf,
+        oci_image: String,
+        services_config: ServicesConfig,
+    ) -> TestEnvironmentConfig<SpinShim> {
+        TestEnvironmentConfig {
+            services_config,
+            create_runtime: Box::new(move |env| {
+                let oci_image = format!(
+                    "localhost:{port}/{oci_image}:latest",
+                    port = env
+                        .get_port(5000)?
+                        .context("environment doesn't expose port for OCI registry")?
+                );
+                SpinShim::registry_push(&spin_binary, &oci_image, env)?;
+                SpinShim::image_pull(&ctr_binary, &oci_image)?;
+                SpinShim::start(&ctr_binary, env, &oci_image, CTR_RUN_ID)
+            }),
+        }
+    }
+
+    pub fn registry_push<R>(
+        spin_binary_path: &Path,
+        image: &str,
+        env: &mut TestEnvironment<R>,
+    ) -> anyhow::Result<()> {
+        let mut cmd = Command::new(spin_binary_path);
+        cmd.args(["registry", "push", "-k"]).arg(image);
+        env.run_in(&mut cmd)
+            .context("failed to push spin app to registry with 'spin'")?;
+        Ok(())
+    }
+
+    pub fn image_pull(ctr_binary_path: &Path, image: &str) -> anyhow::Result<()> {
+        Command::new(ctr_binary_path)
+            .args(["image", "pull"])
+            .arg(image)
+            .output()
+            .context("failed to pull spin app with 'ctr'")?;
+        // TODO: assess output
```

Comment:
Can we address this todo? Just with an 


```suggestion
        anyhow::ensure!(output.status.success(), "pulling image failed");
```